FEATURE: Full width sidebar

keegangeorge opened the pull request:
Added "Full" option for left and right sidebars to appear in full height.

Currently, left sidebar works.

@angusmcleod Do you know why when I check the conditional to see whether currently there is a left or right sidebar, it outputs it as both many times (see image below)? This is resulting in `push-right` and `push-left` being applied to the `#main-outlet` and breaking the design, despite there being a conditional that should prevent it.

```js
  handleFullSidebar() {
    if ($('aside.sidebar.left.full')) {
      $("#main-outlet").addClass("push-right");
      $('.desktop-view').css('overflow-x', "hidden");
      $('aside.sidebar.left').addClass('full-left');
    }    
    if ($('aside.sidebar.right.full')) {
      $("#main-outlet").addClass('push-left');
      $('aside.sidebar.right').addClass('full-right');
    }
  },

```
I've also tried using the variables `hasRightSidebar` and `hasLeftSidebar` but the same issue presents itself.

<img width="643" alt="Screen Shot 2021-07-23 at 2 54 17 PM" src="https://user-images.githubusercontent.com/30090424/126845434-9ec7c14f-81d4-4459-aefc-94555cc94362.png">

keegangeorge committed:
DEV: Add setting to enable full height sidebar

keegangeorge committed:
FEATURE: Setup desktop style for left full height sidebar

keegangeorge committed:
FEATURE: Add code distinction for full right sidebar

angusmcleod commented:
@keegangeorge There could be a few things going on here, but firstly we need to use the existing method of updating the sidebar classes instead of jQuery, i.e. ``buildSidebarClasses``. That will put us in a better position to assess this.

keegangeorge commented:
Thanks @angusmcleod, that works **but** how would I add classes for the `#main-outlet`?

keegangeorge committed:
FIX: Change to not jquery

angusmcleod committed:
IMPROVE: add body class changes to seperate init function and related fixes

keegangeorge committed:
FIX: Full sidebar positioning, content positioning, overflow

keegangeorge commented:
Hey @angusmcleod, I've finished tweaking the full-width sidebar, so it should be ready to be merged after you give a look over, so @MeghnaAJ can begin styling things. Will be moving onto the icon functionality now

keegangeorge committed:
FIX: Ability to add multiple widgets in full sidebar

Previously only first widget can be seen although multiple are present in the sidebar due to full height on widget container. Moved styling and height to sidebar aside instead.

keegangeorge committed:
UX: Add separators for multiple widgets in full position

angusmcleod committed:
Left full position improvements

* Position full requires different left sidebar topic handling
* Prevent right sidebar being used when left sidebar is full
* Ensure classes are removed from the body if full setting no longer applies due to context switching
* Handle header alignment

angusmcleod committed:
Use new computed property for full setting

angusmcleod commented on assets/javascripts/discourse/mixins/sidebars.js.es6:
```diff
@@ -231,7 +263,12 @@ export default Mixin.create({
     }
+    if (hasLeftSidebar && leftFull) {
+      offset = 0;
+      root.style.setProperty('overflow-x', 'hidden');
```
@keegangeorge Two things on this. Firstly, you shouldn't use a computed property callback (i.e. this function) to perform another "set" operation. If you need to set something else as a result of a computed property changing, use a seperate funciton. Always keep computed property functions focused on single type of output, in this case the styles being applied to the main element. 

If you need to perform other behaviour as a result of changes to the same properties observed by ``mainStyle`` set up a seperate computed property or observer to do that.

angusmcleod commented on assets/javascripts/discourse/mixins/sidebars.js.es6:
```diff
@@ -231,7 +263,12 @@ export default Mixin.create({
+      root.style.setProperty('overflow-x', 'hidden');
+    }
     style += `width: ${offset > 0 ? `calc(100% - ${offset}px)` : '100%'}`;
+
```
Avoid formatting unconnected with code changes in a PR.

angusmcleod reviewed (changes requested):
@keegangeorge I've made a few additional changes for the reasons listed in the [commit message of my commit](https://github.com/paviliondev/discourse-layouts/pull/14/commits/872d3baf7a73c11c76ddbd68083d0821625eb210). In addition to those, a few other notes. Once you've reviewed the commit I made, and dealt with these notes, this can be marged. 

Just a minor note about commit messages. We're using the commit prefix convention used by Discourse (see meta topic), so this should be ``FEATURE: Full width left sidebar``. Also we're not currently using github badges.

keegangeorge committed:
IMRPOVE: Move root style to separate computed property
